active/0003-extensible-resource-deployment-observability-improvements.md

```diff
@@ -0,0 +1,249 @@
+
+```diff
+{
+  ...
+  "type": "Microsoft.Resources/deployments",
+  "properties": {
+    "outputResources": [
+      // Azure resource
+      {
+        "id": "/subscriptions/00000000-0000-0000-0000-000000000001/resourceGroups/my-resource-group/providers/Microsoft.Storage/storageAccounts/my-storage-account"
+      },
+      // Extensible resource
+      {
++       "symbolicName": "myService",
++       "resourceType": "core/Service@v1",
++       "deploymentId": "/subscriptions/00000000-0000-0000-0000-000000000001/resourceGroups/my-resource-group/providers/Microsoft.Resources/deployments/my-deployment",
++       "deploymentProvider": "kubernetes"
+      }
+    ]
+  }
+}
+```
+
+### `Microsoft.Resources/deployments/operations` API changes for GET and LIST
+
+#### Updating `targetResource` to include symbolic name and provider for extensible resources
+
+To align with modifications made to the `Microsoft.Resources/deployments API`, the response body now incorporates the addition of `symbolicName` and `deploymentProvider`.
+
+```diff
+{
+  ...
+  "properties": {
+    "provisioningOperation": "Create",
+    "provisioningState": "Succeeded",
+    "targetResource": {
+      "resourceType": "core/Service@v1",
++     "symbolicName": "myService",
++     "deploymentProvider": {
++       "alias": "kubernetes",
++       "name": "Kubernetes",
++       "version": "1.27.8"
++     }
+    }
+  }
+}
+```
+
+### Client side changes
+
+### Azure CLI and Azure PowerShell
+
+Updating the Azure CLI and Azure PowerShell to effectively communicate deployment information for extensible resources is crucial for providing users with clear and actionable insights. A practical approach to this could include enhancing the output with informative log messages that clearly summarize essential details. For instance, implementing a log message format such as:
+
+```
+VERBOSE: 00:00:00 - Extensible resource with symbolic name 'myService' of type 'core/Service@v1' has provisioning status: 'Succeeded'. The resource is provisioned by 'Kubernetes@1.27.8'.
+```
+
+## Drawbacks
+
+### Inconsistency between Azure resources and extensible resources
+
+The new properties are added exclusively for extensible resources, leaving Azure resources unchanged, which avoids introducing breaking changes to the API, SDK, and clients. However, this approach leads to an inconsistency in the schema between Azure resources and extensible resources. Such a disparity could potentially cause confusion, as users might expect a uniform schema across all resource kinds. Clear documentation might be required to manage user expectations and explain the rationale behind the differing schemas.
```

Comment:
You could add the same fields with the same meaning for Azure. Imagine if every output resource defined:

- `symbolicName`
- `resourceType`
- `deploymentId`
- `deploymentProvider`
- (additional *keys* as necessary to identify the resource, for Azure this is `id`)

Reply:
Indeed, this is actaully mentioned in the "Unresolved Questions" section below. For the sake of consistency, implementing this seems appropriate. However, after discussion with other team members yesterday, the decision was made to refrain from implementation at this time. This is due to the fact that integrating the additional properties for Azure resources would necessitate changes across quite a few components of the deployment engine, which carries a considerable risk. Moreover, given that Deployment Stacks might be the sole consumer the `outputResources`, the effort required may not justify the benefits.